### PR TITLE
reshuffles names for blackbox/whitebox contexts, changes bundle notation

### PR DESCRIPTION
**Pull request by xeno-by**
review @retronym

### CODE REVIEW

File: src/compiler/scala/reflect/macros/runtime/JavaReflectionRuntimes.scala [hunk 1]
```diff
@@ -19,7 +19,7 @@ trait JavaReflectionRuntimes {
       macroLogVerbose(s"successfully loaded macro impl as ($implClass, $implMeth)")
       args => {
         val implObj =
-          if (isBundle) implClass.getConstructor(classOf[ApiContext]).newInstance(args.c)
+          if (isBundle) implClass.getConstructors().head.newInstance(args.c)
```
Comment (retronym) [inline src/compiler/scala/reflect/macros/runtime/JavaReflectionRuntimes.scala]: Doesn't handle overloading. For simplicity, you might want to rule out references to macro bundles with overloaded constructors.

Comment (xeno-by) [inline src/compiler/scala/reflect/macros/runtime/JavaReflectionRuntimes.scala]: Nice catch. Overloading is not a problem as long as we have a constructor of a suitable type. Will fix this.

Comment (xeno-by) [inline src/compiler/scala/reflect/macros/runtime/JavaReflectionRuntimes.scala]: Also, the macro bundle checker looks only at the primary constructor, and we can't have overloaded primary constructors :)

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 2]
```diff
@@ -603,32 +599,31 @@ trait Definitions extends api.StandardDefinitions {
...
-      tp.baseClasses.contains(BlackboxMacroClass)
-
-    def isMacroBundleType(tp: Type) = tp.baseClasses match {
-      case _ :: proto :: _ if isMacroBundleProtoType(proto.tpe) => true
-      case _ => false
+    private def macroBundleParamInfo(tp: Type) = {
+      val ctor = tp.erasure.typeSymbol.primaryConstructor
+      ctor.paramss match {
+        case List(List(c)) =>
+          if (c.info != ErrorType && c.info != NoType && c.info != NothingTpe && c.info != NullTpe) NoType
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala]: discarded `if`

Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Definitions.scala]: very embarassing, thank you!

**Comment by retronym** [comment]
PLS SYNCH

**Comment by retronym** [comment]
LGTM. The naming is consistent with the collections library. Getting rid of `compileLate` is massive.

**Comment by xeno-by** [comment]
Can we merge this into master or not yet?

**Comment by retronym** [comment]
Yes, I think this one makes sense to merge pre-M8. Doing so now.

**Comment by retronym** [comment]
Do you support macro bundles in 2.10.x via paradise? If so, it would be great to publish a new version of that to coincide with M8.

**Comment by xeno-by** [comment]
@retronym No, I don't, because they needed `*Macro` classes. But now I very well could. Good idea, thanks!

**Pull request merged**